spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/secretmanager/SecretManagerPropertySource.java

```diff
@@ -16,117 +16,119 @@
 
 package org.springframework.cloud.gcp.autoconfigure.secretmanager;
 
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Map;
-
+import com.google.api.gax.rpc.NotFoundException;
 import com.google.cloud.secretmanager.v1beta1.AccessSecretVersionResponse;
-import com.google.cloud.secretmanager.v1beta1.ProjectName;
-import com.google.cloud.secretmanager.v1beta1.Secret;
 import com.google.cloud.secretmanager.v1beta1.SecretManagerServiceClient;
-import com.google.cloud.secretmanager.v1beta1.SecretManagerServiceClient.ListSecretsPagedResponse;
 import com.google.cloud.secretmanager.v1beta1.SecretVersionName;
 import com.google.protobuf.ByteString;
-import org.apache.commons.logging.Log;
-import org.apache.commons.logging.LogFactory;
 
 import org.springframework.cloud.gcp.core.GcpProjectIdProvider;
 import org.springframework.core.env.EnumerablePropertySource;
 
 /**
- * Retrieves secrets from GCP Secret Manager under the current GCP project.
+ * A property source for Secret Manager which accesses the Secret Manager APIs when {@link #getProperty} is called.
  *
  * @author Daniel Zou
  * @author Eddú Meléndez
  * @since 1.2.2
  */
 public class SecretManagerPropertySource extends EnumerablePropertySource<SecretManagerServiceClient> {
 
-	private static final Log LOGGER = LogFactory.getLog(SecretManagerPropertySource.class);
-
-	private static final String LATEST_VERSION_STRING = "latest";
-
-	private final Map<String, Object> properties;
+	private static final String GCP_SECRET_PREFIX = "gcp-secret/";
```

Comment:
Did the Secrets Manager team comment on this prefix? Have you looked into using `sm://`?

Reply:
Colon can't be used in a property name because it [means something special in SPEL](https://stackoverflow.com/questions/31711308/spring-value-escape-colon-in-default-value). I wouldn't want to try to escape the colon either; I think it would look worse with the back slashes.

Am open to other secret prefixes though.

Reply:
I wonder if we can tap into that "special meaning" and configure it for our purpose.

Reply:
Colon is like the null-safety operator. If the left side of the colon is null, then it defaults to the value on the right side.

Reply:
This is only the case if SPEL is used, such as with `@Value`, right?
Notice that with `@Value` on a `Resource` it's interpreted as a protocol instead.
For example, see the "gs://" implementation we have for Storage.

Reply:
Ok so I found something interesting. It seems like doing the `gcp-sm://` works but has slightly unexpected behavior.

- If the secret exists, then doing the `gcp-sm://path/to/secret` works fine. It will load the secret without issue.

- If the secret property source returns null for any secret property, then it will fallback to the evaluating the left-hand side and if that is null it will show the right hand side. See in our sample, we have secrets named `application-secret`, so in our sample, if the secrets are not present, you see that it will render `//application-secret` instead of the secret value.

![image](https://user-images.githubusercontent.com/3209274/78726721-b1e8b400-7900-11ea-9f42-4f209872a1c8.png)

So I guess if we move forward with throwing an exception in the cases where a secret is missing or the secret URL is not formatted correctly then user will never reach this unexpected behavior.

What are your thoughts?

Reply:
I'm not following fully. Let's discuss.

Reply:
I'll look a little more into to understand if it's fixable.

Just to summarize our discussion this morning, using colon has the consequence where if we ever return `null` for a secret, it will interpret the colon as the null safety operator. This is why it was printing out `//application-secret` for the secret that did not exist, because the original string was `gcp-sm://application-secret`; after resolving it to null, it treated the contents of the right of the colon as the "default value".

Reply:
Ran the code through the debugger, the colon is a special character; i don't think this is configurable. I think we should avoid it to avoid unexpected behavior in the future with null values.

I would be open to other choices of prefixes, maybe `sm|` or `gcp-sm|` these seem to all work.

Reply:
Should we ever return `null` though? Is there something wrong with throwing an exception for a property that doesn't exist?

Reply:
Yeah I can see the argument where if a user specifies a secret using our syntax, there is the expectation that a secret is there; if it's missing then that's most likely an error and they probably don't want null there.

I would be comfortable with either approach. We can still throw exception using `|` too by the way.

To me it seems the tradeoff here is purely cosmetic. I.e. if the cosmetic benefits of using `:` outweighs the risk that we need to change behavior to return `null` someday.

Reply:
What's the prefix that the Secret Manager team suggested we use? There's value in being compatible with other libraries for this product. Let's confirm with the product team.

Reply:
Based on discussion offline, looks like `sm://` has precedent of being used, so I just changed it to that for now.